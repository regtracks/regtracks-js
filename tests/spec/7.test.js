const Track = require('../../index');

describe('tests section 7 of the specification', () => {
  test('section 7.1: variables from collection', () => {
    let schema = String.raw`
    @vars
    ||
    {
      any a to z as str
      ($str)
      }
      times forever
    ||
    `;

    const track = new Track(schema);

    const good = 'aababcabc';
    const bad = 'adada';

    expect(track.test(good, 'vars')).toBe(true);
    expect(track.test(bad, 'vars')).toBe(false);
  });

  test('section 7.2: variables passed to parser', () => {
    let schema = String.raw`
    @vars
    ||
    any a to z
      times forever
        after ($delim)
    ($end)
    ||
    `;

    const track = new Track(schema);
    expect(track.parser.tree.tree).toMatchSnapshot();

    const vars = {
      delim: ',',
      end: '.',
    };

    let good = 'a,b,c,d.';
    let good2 = 'x.';
    let bad = 'e-b-d,';
    let bad2 = 'e,b,d';

    expect(track.test(good, 'vars', {}, vars)).toBe(true);
    expect(track.test(good2, 'vars', {}, vars)).toBe(true);
    expect(track.test(bad, 'vars', {}, vars)).toBe(false);
    expect(track.test(bad2, 'vars', {}, vars)).toBe(false);

    expect(() => track.test(good, 'vars')).toThrow();
  });

  test('section 7.3: variable names', () => {
    let illFormedSchema = String.raw`
      @morevars
      ||
      ($bad!name)
      ||
    `;

    let illFormedSchema2 = String.raw`
      @morevars
      ||
      ($not.allowed)
      ||
    `;

    expect(() => new Track(illFormedSchema)).toThrow();
    expect(() => new Track(illFormedSchema2)).toThrow();
  });
});
