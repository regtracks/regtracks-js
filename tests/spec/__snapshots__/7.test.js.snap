// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests section 7 of the specification section 7.1: variables 1`] = `
Object {
  "vars": Object {
    "name": "vars",
    "root": Object {
      "children": Array [
        Object {
          "collectionIdent": undefined,
          "forks": false,
          "inclusive": true,
          "matches": Array [
            Object {
              "matches": /\\[a-z\\]/,
              "type": "character_range_criterion",
            },
          ],
          "notMatch": false,
          "optional": false,
          "repeat": Object {
            "after": Object {
              "collectionIdent": undefined,
              "forks": false,
              "inclusive": true,
              "matches": Array [
                Object {
                  "matches": "delim",
                  "type": "variable",
                },
              ],
              "notMatch": false,
              "optional": false,
              "repeat": undefined,
              "type": "matching_rule",
            },
            "times": Array [
              1,
              -1,
            ],
            "type": "repeat_modifier",
          },
          "type": "matching_rule",
        },
        Object {
          "collectionIdent": undefined,
          "forks": false,
          "inclusive": true,
          "matches": Array [
            Object {
              "matches": "end",
              "type": "variable",
            },
          ],
          "notMatch": false,
          "optional": false,
          "repeat": undefined,
          "type": "matching_rule",
        },
      ],
      "collectionIdent": undefined,
      "forks": false,
      "notMatch": false,
      "optional": false,
      "parent": undefined,
      "repeat": undefined,
      "type": "block",
    },
    "type": "pattern",
  },
}
`;
